### PR TITLE
SVCB improvements

### LINKED ISSUE #10051: SVCB/HTTPS improvements: Automatic hints
- Program: Authoritative
 - Issue type: Feature request

### Short description
It would be great if one could create an HTTPS/SVCB record where PowerDNS would fill in the hints based on A and AAAA records at the target name.

### Usecase
When updating an IP address, it is easy to forget to update other records (like SVCB and HTTPS) with the same address. Having PowerDNS automatically use in-zone data for these hints prevents this.

### Description
SVCB/HTTPS records should accept `ipv4hint=auto` and `ipv6hint=auto` e.g.

```
example.com.  IN HTTPS 1 www.example.com. ipv4hint=auto ipv6hint=auto
```

When serving out such a record, it should find the appropriate A/AAAA (for www.example.com) and fill in the values in the hints before serving out the record.

When doing an XFR, PowerDNS should always expand these 'auto' fields for compatibility with non (or older) PowerDNS secondaries.

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
This PR adds a bunch of improvements to the SVCB (and derivative) records.

* We now use a parser for "Value Lists" based on the draft's ABNF
* Tests for this parser and the `parseRFC1035CharString` parser are added
* Writing test version of these records is also improved
* There is now a super-class for SVCB-like records. This should allow us to easily add new records based on SVCB
* PowerDNS now supports `ipv{4,6}hints=auto` as parameters (Closes #10051)
  * When set, PowerDNS will lookup the A/AAAA records for the target of the record and supplant this data before serving/XFRing

### To Do
* [x] Tests for auto-hints
* [x] pdnsutil tests for ipv{4,6}hints=auto without corresponding A or AAAA records
* [x] Documentation for auto-hints
* [x] A config-setting for auto-hints?

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by pieterlexis** [comment]
This is ready for review

**Comment by mnordhoff** [comment]
Draft -03 is out. Looks like there may be minor changes to parsing to make it simpler (and possibly make zone-check more complicated).

https://www.ietf.org/rfcdiff?url2=draft-ietf-dnsop-svcb-https-03
https://www.ietf.org/archive/id/draft-ietf-dnsop-svcb-https-03.html
https://www.ietf.org/archive/id/draft-ietf-dnsop-svcb-https-03.html#name-decoding-text-in-zone-files

File: docs/guides/svcb.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
+Using SVCB and derived records
+==============================
+The PowerDNS Authoritative Server has support for the SVCB and derived records (like HTTPS).
```
Comment (Habbie) [inline docs/guides/svcb.rst:3]:
```suggestion
The PowerDNS Authoritative Server has support for the SVCB record and derived records (like HTTPS).
```

File: docs/guides/svcb.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+This support includes doing the standards recommended following of alias-form records in-zone and adding those to the additional section.
+Apart from that, there's the PowerDNS special for "autohints".
+
+Automatic hints
+---------------
+PowerDNS can automatically fill in ``ipv4hint`` and ``ipv6hint`` parameters in SVCB records based on A and AAAA records already in the zones.
```
Comment (Habbie) [inline docs/guides/svcb.rst:9]:
```suggestion
PowerDNS can automatically fill in ``ipv4hint`` and ``ipv6hint`` parameters in SVCB records based on A and AAAA records already present in the zone.
```

File: docs/guides/svcb.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+  www.example.com  IN HTTPS 1 . ipv4hint=auto
+  www.example.com  IN A     192.0.2.1
+  www.example.com  IN AAAA  2001:db8::1
+
+  ipv6.example.com IN HTTPS 1 . ipv6hint=auto
+  ipv6.example.com IN AAAA  2001:db8::2
+  ipv6.example.com IN AAAA  2001:db8::3
+
+A query for example.org|HTTPS will be responded to like this::
```
Comment (Habbie) [inline docs/guides/svcb.rst:22]:
```suggestion
A query for ``HTTPS example.org`` will be responded to like this::
```

File: docs/guides/svcb.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  ;; ANSWER SECTION:
+  example.org.		3600	IN	HTTPS	0 www.example.com
+
+  ;; ADDITIONAL SECTION:
+  www.example.org.	3600	IN	A	192.0.2.1
+  www.example.org.	3600	IN	HTTPS	1 . ipv4hint=192.0.2.1
+  www.example.org.	3600	IN	AAAA	2001:db8::1
+
+Notice that PowerDNS did additional processing and added the target HTTPS record,
+and the address records to the additional section.
```
Comment (Habbie) [inline docs/guides/svcb.rst:36]:
```suggestion
Notice that PowerDNS did additional processing and added the target HTTPS record, and the address records, to the additional section.
```

File: docs/guides/svcb.rst [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+Also notice that the ipv4hint in the www.example.org|HTTPS records is set to the value of the A record.
```
Comment (Habbie) [inline docs/guides/svcb.rst:37]:
```suggestion
Also notice that the ipv4hint in the ``www.example.org HTTPS`` records is set to the value of the A record.
```

File: pdns/dnsrecords.cc [hunk 6]
```diff
@@ -716,6 +716,75 @@ string APLRecordContent::getZoneRepresentation(bool noDot) const {
 
 /* APL end */
 
+/* SVCB start */
+bool SVCBBaseRecordContent::autoHint(const SvcParam::SvcParamKey &key) const {
+  auto p = std::find_if(d_params.begin(), d_params.end(),
+      [&key](const SvcParam &param) {
+        return param.getKey() == key;
+      });
+  if (p == d_params.end()) {
```
Comment (Habbie) [inline pdns/dnsrecords.cc:725]: This bit appears to be repeated quite a few times.

File: pdns/dnsrecords.hh [hunk 7]
```diff
@@ -497,30 +497,40 @@ private:
...
+  public:
+    const DNSName& getTarget() const {return d_target;}
+    uint16_t getPriority() const {return d_priority;}
+    // Returns true if a value for |key| was set to 'auto'
+    bool autoHint(const SvcParam::SvcParamKey &key) const;
+    // Sets the |addresses| to the existing hints for |key|
+    void setHints(const SvcParam::SvcParamKey &key, const std::vector<ComboAddress> &addresses);
+    // Removes the parameter for |key| from d_params
+    void removeParam(const SvcParam::SvcParamKey &key);
+    // Whether or not there are any param
```
Comment (Habbie) [inline pdns/dnsrecords.hh:511]:
```suggestion
    // Whether or not there are any parameters
```

File: pdns/packethandler.cc [hunk 8]
```diff
@@ -538,6 +522,41 @@ void PacketHandler::doAdditionalProcessing(DNSPacket& p, std::unique_ptr<DNSPack
       }
     }
   }
+  // TODO should we have a setting to do this?
```
Comment (Habbie) [inline pdns/packethandler.cc:525]: probably yes

File: pdns/packethandler.cc [hunk 9]
```diff
@@ -538,6 +522,41 @@ void PacketHandler::doAdditionalProcessing(DNSPacket& p, std::unique_ptr<DNSPack
...
+  for (auto &rec : r->getServiceRecords()) {
+    // Process auto hints
+    auto rrc = getRR<SVCBBaseRecordContent>(rec->dr);
+    DNSName target = rrc->getTarget().isRoot() ? rec->dr.d_name : rrc->getTarget();
+    if (rrc->autoHint(SvcParam::ipv4hint)) {
```
Comment (Habbie) [inline pdns/packethandler.cc:530]: I wonder if we could share more code with the AAAA block below.

**Review by Habbie — CHANGES_REQUESTED** [review]
Awesome work! Some comments.

File: pdns/packethandler.cc (hunk 9 above)
Comment (pieterlexis) [reply pdns/packethandler.cc]: I tried to make this better/shorter, but it became worse.

File: docs/guides/svcb.rst [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  ipv6.example.com IN AAAA  2001:db8::2
+  ipv6.example.com IN AAAA  2001:db8::3
+
+A query for ``example.org|HTTPS`` will be responded to like this::
+
+  ;; QUESTION SECTION:
+  ;example.org.			IN	HTTPS
+
+  ;; ANSWER SECTION:
+  example.org.		3600	IN	HTTPS	0 www.example.com
```
Comment (Habbie) [inline docs/guides/svcb.rst:32]: This says `.com` but the next few lines say `.org`

File: docs/guides/svcb.rst [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  www.example.org.	3600	IN	HTTPS	1 . ipv4hint=192.0.2.1
+  www.example.org.	3600	IN	AAAA	2001:db8::1
+
+Notice that PowerDNS did additional processing and added the target HTTPS record, and the address records to the additional section.
+Also notice that the ipv4hint in the ``www.example.org|HTTPS`` records is set to the value of the A record.
+
+PowerDNS takes all the corresponding address records, e.g. when asking for ipv6.example.com::
+
+  ;; QUESTION SECTION:
+  ;ipv6.example.org.		IN	HTTPS
```
Comment (Habbie) [inline docs/guides/svcb.rst:45]: More org/com confusion here.

**Review by Habbie — CHANGES_REQUESTED** [review]
LGTM except for the doc typos!

**Review by Habbie — APPROVED** [review]

**Pull request merged**